### DISCUSSION TITLE
React Native Support

### DISCUSSION THREAD

**Discussion by tsiege** [facebook/lexical#2410] [Q&A] [answered]
I know this is a big ask because it would require Native Bindings in Kotlin and Swift, but are there any plans to get React Native support for Lexical? It would be a game changer for many and to me would be a huge advantage in moving away from ProseMirror to Lexical. The use cases to me are obvious, being able to use one project across web and mobile operating systems

**Comment by amyworrall** [comment]
Which would be the most important considerations for you in a hypothetical React Native version of Lexical?

* Parity with web (supports the same node types, same features etc)
* Ability to extend Lexical via React Native, vs having to write new native code to extend Lexical's functionality
* Collaboration, serialized file format interchange, etc
* Multi-platform support in RN (i.e. would it only be useful to you if we had both iOS and Android, or would an iOS-alone RN version be of use?)

**Comment by tsiege** [comment]
Hey @amyworrall. Thanks for the quick reply. I think it's a combination of 1,2, and 4. I'm not sure how we could use Lexical outside of a web view on RN, which is problematic. Ideally there would be a Lexcial framework that is as low level as Lexical, but allows for native bindings to iOS and Android. I'd have to dig into this more but it seems likely that the core engine, such as `editorState`, of Lexical could be designed so that it's platform agnostic and exposes an API that swift or kotlin could use to render the WYSIWYG

**Comment by CanRau** [comment]
Yes please 🥺 React Native support would be tremendous 🥳

Ideally all 4 points 😅😇

**Comment by amyworrall** [comment, marked as answer]
As Dominic has hinted elsewhere, I’m leading a project to build native Lexical on iOS. I’m in the very early days of looking in to what a React Native wrapper around Lexical iOS would look like. The main sticking point is ensuring synchronous callbacks from JS to native, which requires bleeding edge RN features, but aside from that the issue will be the completely different layout engine for TextKit on iOS compared with DOM/CSS. So even if we enable building node subclasses in JS under RN, those node subclasses will have to use iOS semantics (providing string attributes rather than Dom nodes). That’s why I’m trying to find out what the most useful level of RN support would be, to ensure we focus on making that a good experience first. 

(I know I’ve not mentioned Android either. We have nothing to talk about there yet, alas.)

**Reply by natew** [reply]
@amyworrall just curious if there is still movement on this?

**Reply by stephentuso** [reply]
Haven't tested myself yet, but now an option is to `'use dom'` the normal react version

**Reply by CanRau** [reply]
yes very curious if anyone's already tried it with lexical?

**Reply by flakis05** [reply]
@stephentuso, @natew, @CanRau
Did any of you tried “use dom” with lexical?

**Reply by CanRau** [reply]
@flakis05 unfortunately still didn't get the chance 🥲
Tho I'm also very curious 🤓

**Reply by Nantris** [reply]
My understanding is that `'use dom'` just makes a webview under the hood? And if it did anything else I severely doubt it would work, but I'd be very interested to see it tested.

**Reply by grifotv** [reply]
Lexical in React Native with DOM components https://www.youtube.com/watch?v=CxORa1tXMjw

**Reply by Nantris** [reply]
@grifotv it's just a webview. https://docs.expo.dev/guides/dom-components/#webview-props

**Reply by jball-117** [reply]
That video is nice but I get "Metro error: await is not defined" on web. Anyone have any luck?

**Comment by thegreatercurve** [comment]
Closing and moving to a discussion as question above has been answered.

**Comment by strdr4605** [comment]
So, there is no intent to port lexical to ReactNative in a Webview? 
What is the intended MVP to be delivered?
Will it be a simple Editor with lexical API, maybe WYSIWYG Editor? 
Are there any plans for Collaborative editing?
@amyworrall

**Comment by pro-vi** [comment]
Bumping this - would love an update on the progress if available!

**Comment by phillipb** [comment]
@amyworrall I'd love to know if the iOS version is still happening. I actually do NOT require react native support. I'd be happy with an iOS version that could read the editorState from the web and render identical looking nodes. Any update on the project would be super helpful. Thank you all for a great project!

**Comment by strdr4605** [comment]
Hey @zurfyx, @acywatson, what is the current status of React Native support? Maybe you can answer these questions https://github.com/facebook/lexical/discussions/2410#discussioncomment-3000714

**Reply by acywatson** [reply]
Hey @strdr4605 (and others) - you might be interested in this:  https://github.com/facebook/lexical/pull/4124

We're pretty far from a guarantee that anything will be implemented, but we are starting to contemplate changes to the core to accommodate the different reconciliation strategies that would be required for React Native support.

**Reply by alidcast** [reply]
> We're pretty far from a guarantee that anything will be implemented

as in, unsure of whether they'll be RN support or of which direction it'll take?

**Reply by acywatson** [reply]
Both, honestly. We're in early stages, but we are moving forward with experimentation. I just don't want to make any promises we can't keep.

**Reply by Nantris** [reply]
@acywatson is it still the plan to only experiment with iOS native support? Is there any plan to experiment with an Android native version as well?

**Reply by acywatson** [reply]
We did end up open-sourcing Lexical iOS:

https://github.com/facebook/lexical-ios

and we also have an extremely basic android version, but I am still not confident about that or RN support being released into OSS in the near future, unfortunately. @amyworrall

**Reply by alidcast** [reply]
Having the ios renderer out is promising enough! I imagine the RN layer would naturally come last anyway, after establishing platform compatibility.

Thanks for the transparent update.

**Reply by YohanSciubukgian** [reply]
@acywatson Do you have any update to share on the RN support for cross IOS & Android support please?

**Reply by username14415** [reply]
@acywatson any progress of native android version?

**Comment by strdr4605** [comment]
Has anyone tried lexical in a webview with React Native? 
We plan to migrate our mobile editor in the next few months but most probably lexical for React Native will not be available soon.
We did a POC and it seems to work (at least as a temporary solution).

**Reply by CanRau** [reply]
I found a couple people in the Discord doing it and at least 1 seemed pretty happy

**Reply by Guneetgstar** [reply]
@CanRau Can you share the reference, please? Have used `lexical` for `react` but now there is a requirement to build `react-native` apps as well. A sample code with lexical inside `react-native-webview` would be much appreciated.

**Reply by strdr4605** [reply]
@Guneetgstar, I might have a repo or article written on this topic by mid-January.

**Reply by CanRau** [reply]
@Guneetgstar just search webview in lexicals discord

Here's a few

https://discord.com/channels/953974421008293909/964928369793843230/1147618061994709042

https://discord.com/channels/953974421008293909/964928369793843230/1147946700984569947

https://discord.com/channels/953974421008293909/1017916026249281627/1017918375046938696

**Reply by davidfers** [reply]
> @Guneetgstar, I might have a repo or article written on this topic by mid-January.

Hey do you have something on this? It would be super helpful!

**Reply by strdr4605** [reply]
> Hey do you have something on this? It would be super helpful!
> 
> 

Right now fixing some issues with focus and scroll on multi line text with keyboard open. 
But will write about setting up lexical with webview next week probably.

**Reply by strdr4605** [reply]
Hey, I finally managed to finish the guide: https://strdr4605.com/how-to-set-up-lexical-editor-in-react-native
Let me know if you have questions😊

**Reply by acywatson** [reply]
Amazing. @strdr4605, can we add the content to the docs site with a link back to your site?

PR welcome if you want to do that, or we can later.

**Reply by strdr4605** [reply]
@acywatson let's do this a bit later, the article is still fresh so I might add something missing

**Reply by strdr4605** [reply]
@acywatson, do you want to add the whole guide to https://lexical.dev/ or just a section that describes the current state of React Native support, and add the link to my guide?

**Reply by acywatson** [reply]
I was thinking add the content, and also a link - just so people could search the content on the site. 

I'm honestly good either way, though.

**Comment by Hudson-AmalembaL** [comment]
Hello, very nice work on the Lexical framework real life saver.
My question however is this, suppose i have saved the editors state in my database which i can correctly render on my webapplication, in the case that i want to extend my application so that i  have a mobile application that renders the same content in the same format, using react native, is it still posible to render that same stored "lexical editor data" in a react native application?

Knowing this will help a great deal if i can choose to use the lexical framework in my project or not.

Many Thanks.

**Reply by strdr4605** [reply]
You can parse the lexical state in React Native and render it using Text with styles. 

You can even create an adapter for a basic Input.

In our project we created an adapter from/to lexical state with https://github.com/dabakovich/react-native-controlled-mentions

But if you have a more complex state with custom nodes and plugins, for now you can use a webview (still not bulletproof, but it's working, we use it in production), check https://strdr4605.com/how-to-set-up-lexical-editor-in-react-native

**Reply by Hudson-AmalembaL** [reply]
@strdr4605  I've checked the repo, Very much appreciated.

**Reply by davevilela** [reply]
hey @strdr4605 can you share how you converted react-native-controlled-mentions value to lexical state?

**Reply by strdr4605** [reply]
@davevilela, can't help you much here but you can check a fast description of the workflow:
https://gist.github.com/strdr4605/fada2a3bd0d2a05690a475a23f18c404

**Comment by username14415** [comment]
Still no plans for React Native support? Just asking
